crates/ruff_db/src/diagnostic/render.rs

```diff
@@ -17,20 +17,31 @@ use super::{
     Annotation, Diagnostic, DiagnosticFormat, DisplayDiagnosticConfig, Severity, SubDiagnostic,
 };
 
+/// A type that implements `std::fmt::Display` for diagnostic rendering.
+///
+/// It is created via [`Diagnostic::display`].
+///
+/// The lifetime parameters are:
+///
+/// * `'c` is the lifetime of the rendering configuration.
+/// * `'r` is the lifetime of the resolver used to load the contents of `Span`
+///   values. When using Salsa, this most commonly corresponds to the lifetime
+///   of a Salsa `Db`.
+/// * `'d` is the lifetime of the diagnostic being rendered.
 #[derive(Debug)]
-pub(crate) struct DisplayDiagnostic<'a> {
-    config: &'a DisplayDiagnosticConfig,
-    resolver: &'a FileResolver<'a>,
+pub struct DisplayDiagnostic<'c, 'r, 'd> {
```

Comment:
I'm sure you thought about this but are the fine grained lifetimes required or could we just use one saying: You can display this as long as neither `self`, `db` or `config` gets dropped.

Reply:
Yes I did. When it's crate internal, it generally shouldn't matter, because these are all shared borrows and thus subtyping comes into play. And if it does matter, well, then I guess you just fix the lifetimes.

It _can_ matter though. See [this](https://github.com/rust-lang/regex/issues/168) as an example. It's not totally obvious to me that it _will_ matter, but usually when I expose public APIs, I try not to collapse lifetimes. So I mostly just did this by reflex.

Reply:
It's great to see that you have the same ambition for `ruff_db` as for the `regex` crate 😆 

Isn't this slightly different in the sense that `DisplayDiagnostic` never returns any values? I do see the point of having all the lifetimes if `DisplayDiagnostic` had field accessors where it matters that they return the right lifetime. But here it feels like it's mainly: You can only use it as long as the three borrows are valid. 

Either way, not that important. It's just very scary looking

Reply:
Yeah that's what I meant by it might not mattering. As in, I can't easily come up with an example where it matters.

Happy to change it back, but I honestly don't feel strongly.